File: server/src/main/java/rdp/DataSocket.java
package rdp;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.sql.SQLException;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.Queue;

import ttp.FlagType;
import ttp.Processor;
import ttp.Server;
import ttp.TTPPacket;
import util.Utility;

public class DataSocket extends Thread {
	
	// The list of pending datagrams waiting on ACKs
	private LinkedList<PendingDatagram> pendingSentDatagrams;
	
	// The list of datagrams still waiting to be sent
	private Queue<RDPDatagram> datagramSendList;
	
	// The list of sequence numbers which still need to be ACKed
	private Queue<Integer> receivedSequences;
	
	// The list of fragmented packets, yet to be put together
	private HashMap<Integer, FragmentedDatagram> fragmentedData;
	
	// The current send window (number of datagrams to send in the current cycle)
	private int sendWindow;
	
	// The index of the current timeout datagram (the datagram being watched to timeout)
	private int timeoutDatagramIndex;
	
	// The current timeout time in milliseconds
	private double timeoutTime;
	
	// The number of timeouts that have occurred for a send window
	private int numTimeouts;
	
	// The maximum number of allowed pending datagrams
	private final int MAXPENDINGDATAGRAMS = 210;
	
	// The maximum size of a datagram's data
	private final int MAXDATAGRAMDATASIZE = 504 - 16;
	
	// The ssthresh to where the send windows stops growing as fast
	private int ssthresh = MAXPENDINGDATAGRAMS / 2;
	
	// Alpha and beta used in timeout calculations
	private final double ALPHA = .125;
	private final double BETA = .25;
	
	// Estimate and dev used for timeout calculation
	private double estimateRTT = 0;
	private double devRTT = 0;
	
	// The index of the datagram to watch for recalculating current timeout
	private int timeoutWatchIndex;
	
	// The processor used for reading packets
	private final Processor reader;
	
	// The UDP socket for sending and receiving datagrams
	private DatagramSocket socket;
	
	// The current address and port to send to
	private InetAddress sendAddress;
	private int sendPort;
	
	// Whether the thread shoould keep running
	private boolean running = true;
	
	public void setConnection(InetAddress address, int port) {
		this.sendAddress = address;
		this.sendPort = port;
	}
	
	public InetAddress getAddress() {
		return this.sendAddress;
	}
	
	public int getPort() {
		return this.sendPort;
	}
	
	public DataSocket(Processor reader) throws SocketException {
		this.pendingSentDatagrams = new LinkedList<>();
		this.datagramSendList = new LinkedList<>();
		this.receivedSequences = new LinkedList<>();
		this.fragmentedData = new HashMap<>();
		this.sendWindow = 1;
		this.timeoutTime = 1000;
		this.reader = reader;
		this.socket = new DatagramSocket(this.reader.getPort());
		this.socket.setSoTimeout(100);
	}
	
	// The main protocol method
	@Override
	public void run() {
		while (this.running) {
			if (this.pendingSentDatagrams.isEmpty()) {
				try {
					sendProtocol();
					if (this.sendWindow < this.ssthresh) this.sendWindow *= 2;
					else this.sendWindow += 1;
					if (this.sendWindow > this.MAXPENDINGDATAGRAMS) this.sendWindow = this.MAXPENDINGDATAGRAMS;
				} catch (IOException e) {
					timeout();
				}
			} else if (new Date().getTime() - this.pendingSentDatagrams.get(this.timeoutDatagramIndex).timeSent >= this.timeoutTime) timeout();
			while (this.running) {
				try {
					RDPDatagram rdp = receive();
					if (rdp == null) break;
					else {
						if (rdp.getAcknowledgement() > 0) processAck(rdp.getAcknowledgement(), rdp.getSequence());
						if (rdp.getLength() > 0) {
							TTPPacket ttp = rdp.getTTPPacket();
							if (ttp.getFlag() == FlagType.CON) this.receivedSequences.add(-rdp.getSequence());
							else this.receivedSequences.add(rdp.getSequence());
							if (rdp.getHead() > 0) deFragmentPacket(rdp);
							else {
								this.reader.receive(ttp);
								if (this.reader instanceof Server) break;
							}
						}
					}
				} catch (IOException e) {
					timeout();
				}
			}
		}
	}
	
	// Put together fragmented packets
	private void deFragmentPacket(RDPDatagram rdp) {
		if (this.fragmentedData.containsKey(rdp.getHead())) {
			this.fragmentedData.get(rdp.getHead()).addDatagram(rdp);
			if (this.fragmentedData.get(rdp.getHead()).numDatagramsNeeded == 0) this.reader.receive(this.fragmentedData.get(rdp.getHead()).getTTPPacket());
		} else {
			this.fragmentedData.put(rdp.getHead(), new FragmentedDatagram(rdp.getHead(), rdp.getTail()));
			this.fragmentedData.get(rdp.getHead()).addDatagram(rdp);
		}
	}
	
	// When a timeout occurs
	private void timeout() {
		this.sendWindow = 1;
		this.ssthresh /= 2;
		this.numTimeouts++;
		if (this.numTimeouts == 3) {
			int check = this.reader.reconnect();
			if (check == 1) close();
			else if (check == -1) reset();
		}
	}
	
	// Proceses an ACK for the send list
	private void processAck(int ackNumber, int seqNumber) {
		Iterator<PendingDatagram> i = this.pendingSentDatagrams.iterator();
		int index = 0;
		while (i.hasNext()) {
			PendingDatagram pd = i.next();
			if (pd.datagram.getSequence() == ackNumber) {
				i.remove();
				if (pd.datagram.getTTPPacket().getFlag() == FlagType.CON) this.receivedSequences.add(seqNumber);
				if (index == this.timeoutDatagramIndex) this.timeoutDatagramIndex++;
				if (this.timeoutDatagramIndex >= this.pendingSentDatagrams.size()) this.timeoutDatagramIndex = 0;
				if (index == this.timeoutWatchIndex) calculateTimeout(new Date().getTime() - pd.timeSent);
				index = 0;
				i = this.pendingSentDatagrams.iterator();
				while (i.hasNext()) {
					PendingDatagram p = i.next();
					if (p.datagram.getSequence() < ackNumber) {
						p.dupAck++;
						if (p.dupAck == 3) {
							resend(p);
							if (index == this.timeoutDatagramIndex) this.timeoutDatagramIndex++;
							if (this.timeoutDatagramIndex == this.pendingSentDatagrams.size()) this.timeoutDatagramIndex = 0;
							if (index == this.timeoutWatchIndex) this.timeoutWatchIndex++;
						}
						index++;
					} else break;
				}
				break;
			}
			index++;
		}
	}
	
	// Recalculate the current timeout time
	private void calculateTimeout(long sample) {
		this.estimateRTT = (1 - this.ALPHA) * this.estimateRTT + this.ALPHA * sample;
		this.devRTT = (1 - this.BETA) * this.devRTT + this.BETA * Math.abs(sample - this.estimateRTT);
		this.timeoutTime = this.estimateRTT + 4 * this.devRTT;
	}
	
	// Receive the next datagram or return null if there isn't one to receive
	private RDPDatagram receive() throws IOException {
		byte[] buffer = new byte[512];
		DatagramPacket d = new DatagramPacket(buffer, buffer.length);
		try {
			this.socket.receive(d);
			this.sendAddress = d.getAddress();
			this.sendPort = d.getPort();
			RDPDatagram r = new RDPDatagram(d.getData());
			System.out.println(this.reader.getPort());
			System.out.println(r);
			return r;
		} catch (SocketTimeoutException e) {
			return null;
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
	}
	
	// Send the current send window
	private void sendProtocol() throws IOException {
		Date d = new Date();
		int numSent = 0;
		this.timeoutDatagramIndex = 0;
		this.timeoutWatchIndex = 0;
		while (!this.datagramSendList.isEmpty() && numSent < this.sendWindow) {
			RDPDatagram r = this.datagramSendList.poll();
			send(r);
			this.numTimeouts = 0;
			this.pendingSentDatagrams.add(new PendingDatagram(r, 0, d.getTime()));
			numSent++;
		}
		while (numSent < this.sendWindow && !this.receivedSequences.isEmpty()) {
			int seq = this.receivedSequences.poll();
			int realSeq = Math.abs(seq);
			RDPDatagram r = new RDPDatagram(realSeq, 0, 0, null);
			send(r);
			if (seq < 0) this.pendingSentDatagrams.add(new PendingDatagram(r, 0, d.getTime()));
			numSent++;
		}
	}
	
	// Send a datagram
	private void send(RDPDatagram rdp) throws IOException {
		byte[] bytes = rdp.getBytes();
		System.out.println("sending");
		System.out.println(rdp);
		System.out.println(this.sendPort + " " + this.sendAddress);
		this.socket.send(new DatagramPacket(bytes, bytes.length, this.sendAddress, this.sendPort));
	}
	
	// Queue up a datagram to be sent
	public void addSend(TTPPacket ttp) {
		byte[] bytes = ttp.getBytes();
		if (bytes.length > this.MAXDATAGRAMDATASIZE) {
			byte[][] data = Utility.splitBytesData(bytes, this.MAXDATAGRAMDATASIZE - 1);
			RDPDatagram.resetSequenceIfLarger(data.length);
			for (int i = 0; i < data.length; i++) {
				if (i > 0) data[i] = Utility.mergeBytes(new byte[] {bytes[0]}, data[i]);
				TTPPacket t = new TTPPacket(data[i]);
				Integer ack = this.receivedSequences.poll();
				if (ack == null) ack = 0;
				else ack = Math.abs(ack.intValue());
				RDPDatagram r = new RDPDatagram(ack, data.length, t);
				this.datagramSendList.add(r);
			}
		} else {
			Integer ack = this.receivedSequences.poll();
			if (ack == null) ack = 0;
			this.datagramSendList.add(new RDPDatagram(ack, 0, 0, ttp));
		}
	}
	
	// Resend a datagram
	private void resend(PendingDatagram pd) {
		pd.dupAck = 0;
		this.sendWindow /= 2;
		pd.timeSent = new Date().getTime();
		while (this.running) {
			try {
				send(pd.datagram);
				break;
			} catch (IOException e) {
				timeout();
				try {
					this.wait(100);
				} catch (InterruptedException e1) {
					e1.printStackTrace();
				}
			}
		}
	}
	
	// Reset all the sending and receiving
	public void reset() {
		this.datagramSendList.clear();
		this.fragmentedData.clear();
		this.numTimeouts = 0;
		this.pendingSentDatagrams.clear();
		this.receivedSequences.clear();
		this.ssthresh = 210;
		this.timeoutTime = 1000;
	}
	
	// Close the connection
	public void close() {
		this.socket.close();
		try {
			this.reader.close();
		} catch (SQLException e) {
			e.printStackTrace();
		}
		this.running = false;
	}
	
	public Processor getReader() {
		return this.reader;
	}
	
	// Pending datagrams to keep track of duplicate ACKs and timestamps
	private class PendingDatagram {
		private RDPDatagram datagram;
		private int dupAck;
		private long timeSent;
		
		public PendingDatagram(RDPDatagram datagram, int dupAck, long timeSent) {
			this.datagram = datagram;
			this.dupAck = dupAck;
			this.timeSent = timeSent;
		}
	}
	
	// Framented packets to build them fully
	private class FragmentedDatagram {
		private int numDatagramsNeeded;
		private byte[] data;
		
		public FragmentedDatagram(int start, int end) {
			this.numDatagramsNeeded = end - start + 1;
			this.data = new byte[this.numDatagramsNeeded * MAXDATAGRAMDATASIZE];
		}
		
		public void addDatagram(RDPDatagram rdp) {
			this.numDatagramsNeeded--;
			byte[] bytes = rdp.getData();
			if (rdp.getSequence() == rdp.getTail() && bytes.length < MAXDATAGRAMDATASIZE) {
				byte[] temp = new byte[this.data.length - (MAXDATAGRAMDATASIZE - bytes.length)];
				for (int i = 0; i < temp.length; i++) {
					temp[i] = this.data[i];
				}
				this.data = temp;
			}
			for (int i = 0; i < bytes.length; i++) {
				this.data[i + (rdp.getSequence() - rdp.getHead()) * MAXDATAGRAMDATASIZE] = bytes[i];
			}
		}
		
		public TTPPacket getTTPPacket() {
			
			return new TTPPacket(this.data);
		}
	}

}

File: python_client/DataSocket.py
from PyQt5.QtCore import QThread
from PyQt5.QtNetwork import QUdpSocket, QHostAddress
from datetime import datetime as dt
from collections import deque
from RDP_datagram import RDP_Datagram, reset_sequence_if_larger
import TTP_packet as ttp
import utility

MAX_PENDING_DATAGRAMS = 210
MAX_DATAGRAM_SIZE = 504-16
# HOSTADDRESS = QHostAddress('127.0.0.1')
HOSTADDRESS = QHostAddress('10.101.232.30')


class DataSocket(QThread):

    _alpha = .125
    _beta = .25
    _estimateRTT = 0
    _devRTT = 0

    def __init__(self, reader):
        self.serverport = 11000
        QThread.__init__(self)
        self._ssthresh = MAX_PENDING_DATAGRAMS / 2
        self._pending_sent_datagrams = []
        self._datagram_send_list = deque([])
        self._received_sequences = deque([])
        self._fragmented_data = dict()
        self._running = True
        self._send_window = 1
        self.timeout_time = 1000
        self.numtimeouts = 0
        self.timeout_datagram_index = 0
        self.socket = QUdpSocket()
        self.socket.bind()
        self.reader = reader

    def run(self):
        while (self._running):
            if (len(self._pending_sent_datagrams) == 0):
                try:
                    self._send_protocol()
                    if (self._send_window < self._ssthresh):
                        self._send_window *= 2
                    else:
                        self._send_window += 1
                    if (self._send_window > MAX_PENDING_DATAGRAMS):
                        self._send_window = MAX_PENDING_DATAGRAMS
                except Exception as e:
                    self.timeout()
            elif (round(dt.now().microsecond / 1000) - self._pending_sent_datagrams[self.timeout_datagram_index].time_sent >= self.timeout_time):
                self.timeout()
            while (self._running):
                try:
                    rdp = self.receive()
                    if (rdp is None):
                        break
                    else:
                        if (rdp.acknowledgement != 0):
                            self.process_ack(rdp.acknowledgement, rdp.sequence)
                        if len(rdp.data) > 0:
                            self._received_sequences.append(rdp.sequence)
                            if (rdp.head > 0):
                                self.defragment_packet(rdp)
                            else:
                                self.reader.receive(rdp.get_TTP_packet())
                except Exception as e:
                    self.timeout()

    def defragment_packet(self, rdp):
        if (rdp.head in self._fragmented_data):
            self._fragmented_data[rdp.head].add_datagram(rdp)
            if (self._fragmented_data[rdp.head].num_datagrams_needed == 0):
                self.reader.receive(self._fragmented_data[rdp.head].get_TTP_packet())
        else:
            self._fragmented_data[rdp.head] = FragmentedDatagram(rdp.head, rdp.tail)
            self._fragmented_data[rdp.head].add_datagram(rdp)

    def timeout(self):
        self._send_window = 1
        self._ssthresh /= 2
        self._ssthresh = int(self._ssthresh)
        self.numtimeouts += 1
        if (self.numtimeouts == 3):
            check = self.reader.reconnect()
            if (check == 1):
                self.close()
            elif (check == -1):
                self.reset()

    def process_ack(self, ack_number, seq_number):
        index = 0
        for pending_gram in self._pending_sent_datagrams:
            if (pending_gram.datagram.sequence == abs(ack_number)):
                self._pending_sent_datagrams.remove(pending_gram)
                if (index == self.timeout_datagram_index):
                    self.timeout_datagram_index += 1
                if (self.timeout_datagram_index >= len(self._pending_sent_datagrams)):
                    self.timeout_datagram_index = 0
                if (index == self.timeout_watch_index):
                    self.calculate_timeout(round(dt.now().microsecond / 1000) - pending_gram.time_sent)
                index = 0
                for pending_gram in self._pending_sent_datagrams:
                    if(pending_gram.datagram.sequence < ack_number):
                        pending_gram.dup_ack += 1
                        if (pending_gram.dup_ack == 3):
                            self.resend(pending_gram)
                            if (index == self.timeout_datagram_index):
                                self.timeout_datagram_index += 1
                            if (self.timeout_datagram_index == len(self._pending_sent_datagrams)):
                                self.timeout_datagram_index = 0
                            if (index == self.timeout_watch_index):
                                self.timeout_watch_index += 1
                        index += 1
                    else:
                        break
                break
            index += 1

    def calculate_timeout(self, sample):
        self._estimateRTT = (1 - self._alpha) * self._estimateRTT + self._alpha * sample
        self._devRTT = (1 - self._beta) * self._devRTT + self._beta * abs(sample - self._estimateRTT)
        self.timeout_time = self._estimateRTT + 4 * self._devRTT

    def receive(self):
        # self.socket.waitForReadyRead()
        # print('hello world')
        # print(self.socket.readDatagram(512)[0])
        if not self.socket.waitForReadyRead(100):
            return None
        else:
            buffer = self.socket.readDatagram(512)[0]
            return RDP_Datagram(buffer)

    def _send_protocol(self):  # send the current window
        numsent = 0
        self.timeout_datagram_index = 0
        self.timeout_watch_index = 0
        while (len(self._datagram_send_list) > 0 and numsent < self._send_window):
            r = self._datagram_send_list.popleft()
            self.send(r)
            self.numtimeouts = 0
            self._pending_sent_datagrams.append(PendingDatagram(r, 0, round(dt.now().microsecond / 1000)))
            numsent += 1
        while (numsent < self._send_window and len(self._received_sequences) > 0):
            seq = self._received_sequences.popleft()
            real_seq = abs(seq)
            r = RDP_Datagram(acknowledgement=real_seq)
            self.send(r)
            if seq < 0:
                self._pending_sent_datagrams.append(PendingDatagram(r, 0, round(dt.now().microsecond / 1000)))

    def send(self, datagram):  # send a datagram
        bites = datagram.get_bytes()
        print(bites)
        self.socket.writeDatagram(bites, HOSTADDRESS, self.serverport)

    def add_send(self, ttppacket):
        bites = ttppacket.get_bytes()
        if (len(bites) > MAX_DATAGRAM_SIZE):
            data = utility.split_bytes_data(bites, MAX_DATAGRAM_SIZE - 1)
            start = 0
            end = 0
            reset_sequence_if_larger(len(data))
            for i in range(len(data)):
                if i > 0:
                    data[i] = bites[0:1] + data[i]
                t = ttp.TTP_packet(bytes_arr=data[i])
                ack = 0
                if len(self._received_sequences) > 0:
                    ack = self._received_sequences.popleft()
                if i == 0:
                    r = RDP_Datagram(acknowledgement=ack, frag_distance=len(data) - 1, packet=t)
                    start = r.head
                    end = r.tail
                else:
                    r = RDP_Datagram(acknowledgement=ack, head=start, tail=end, packet=t)
                self._datagram_send_list.append(r)
        else:
            ack = 0
            if len(self._received_sequences) > 0:
                ack = self._received_sequences.popleft()
            self._datagram_send_list.append(RDP_Datagram(acknowledgement=ack, packet=ttppacket))

    def resend(self, pd):
        pd.dup_ack = 0
        self._send_window /= 2
        self._send_window = int(self._send_window)
        pd.time_sent = round(dt.now().microsecond / 1000)
        while (self._running):
            try:
                self.send(pd.datagram)
                break
            except Exception as e:
                self.timeout()
                try:
                    self.wait(100)
                except Exception as e:
                    raise e

    def reset(self):
        self._datagram_send_list.clear()
        self._fragmented_data.clear()
        self.numtimeouts = 0
        self._pending_sent_datagrams.clear()
        self._received_sequences.clear()
        self._ssthresh = 210
        self.timeout_time = 1000

    def close(self):
        self.socket.close()
        try:
            self.reader.close()
        except Exception as e:
            raise e
        self._running = False


class PendingDatagram():
    def __init__(self, datagram, dup_ack, time_sent):
        self.datagram = datagram
        self.dup_ack = dup_ack
        self.time_sent = time_sent


class FragmentedDatagram():
    def __init__(self, start, end):
        self.num_datagrams_needed = end - start
        self.data = bytearray(self.num_datagrams_needed * MAX_DATAGRAM_SIZE)

    def add_datagram(self, rdp):
        self.num_datagrams_needed -= 1
        bites = rdp.get_data()
        if (rdp.sequence == rdp.tail and len(bites) < MAX_DATAGRAM_SIZE):
            data_len = len(self.data) - (MAX_DATAGRAM_SIZE - len(bites))
            self.data = self.data[:data_len]
        num = rdp.sequence - rdp.head
        self.data[num:num + len(bites)] = bites

    def get_TTP_packet(self):
        return ttp.TTP_packet(self.data)
